cp3pt0-deployment/migrate_tenant.sh

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env bash
+
+# Licensed Materials - Property of IBM
+# Copyright IBM Corporation 2023. All Rights Reserved
+# US Government Users Restricted Rights -
+# Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
+#
+# This is an internal component, bundled with an official IBM product. 
+# Please refer to that particular license for additional information. 
+
+# ---------- Command arguments ----------
+
+OC=oc
+YQ=yq
+OPERATOR_NS=""
+SERVICES_NS=""
+NS_LIST=""
+CONTROL_NS=""
+CHANNEL="v4.0"
+SOURCE="opencloud-operators"
+CERT_MANAGER_SOURCE="ibm-cert-manager-operator-catalog"
+LICENSING_SOURCE="ibm-licensing-catalog"
+SOURCE_NS="openshift-marketplace"
+INSTALL_MODE="Automatic"
+ENABLE_CERTMANAGER=0
+ENABLE_LICENSING=0
+ENABLE_PRIVATE_CATALOG=0
+NEW_MAPPING=""
+NEW_TENANT=0
+DEBUG=0
+PREVIEW_MODE=1
+
+# ---------- Command variables ----------
+
+# script base directory
+BASE_DIR=$(cd $(dirname "$0")/$(dirname "$(readlink $0)") && pwd -P)
+
+# counter to keep track of installation steps
+STEP=0
+
+# ---------- Main functions ----------
+
+. ${BASE_DIR}/common/utils.sh
+
+function main() {
+    parse_arguments "$@"
+    pre_req
+    
+    # TODO check Cloud Pak compatibility
+
+
+    # # Update common-serivce-maps ConfigMap
+    # ${BASE_DIR}/common/mapping_tenant.sh --operator-namespace $OPERATOR_NS --services-namespace $SERVICES_NS --tethered-namespaces $TETHERED_NS --control-namespace $CONTROL_NS
+    
+    # Scale down CS and ODLM
+    # It helps to prevent re-installing licensing and cert-manager services
+    scale_down_cs
+
+    # Migrate CP2 cluster singleton services
+    local arguments="--enable-cert-manager"
+
+    # Do not migrate Licensing if it is multi-instance enabled mode
+    if [[ "$CONTROL_NS" == "$OPERATOR_NS" ]]; then
+        arguments+=" --enable-licensing"
+    fi
+    
+    if [[ $ENABLE_PRIVATE_CATALOG -eq 1 ]]; then
+        arguments+=" --enable-private-catalog"
+    fi
+    ${BASE_DIR}/migrate_singleton.sh "--operator-namespace" "$OPERATOR_NS" "-c" "$CHANNEL" "--cert-manager-source" "$CERT_MANAGER_SOURCE" "--licensing-source" "$LICENSING_SOURCE" "$arguments"
+
+
+    # Update CommonService CR with OPERATOR_NS and SERVICES_NS
+    # Propogate CommonService CR to every namespace in the tenant
+    update_cscr "$OPERATOR_NS" "$SERVICES_NS" "$NS_LIST"
+
+    # Update ibm-common-service-operator channel
+    for ns in ${NS_LIST//,/ }; do
+        if [ $ENABLE_PRIVATE_CATALOG -eq 0 ]; then
+            update_operator_channel ibm-common-service-operator $ns $CHANNEL $SOURCE $SOURCE_NS $INSTALL_MODE
+        else
+            update_operator_channel ibm-common-service-operator $ns $CHANNEL $SOURCE $ns $INSTALL_MODE
+        fi
+    done
+
+    # wait for operator upgrade
+    # TODO: wait for operator by checking installedCSV in sub
+    sleep 60
+    wait_for_operator "$OPERATOR_NS" "ibm-common-service-operator"
```

Comment:
Can we wait for subscription.status.state == AtLatestKnown instead of waiting for operator, so that the `sleep` can be removed?